File: usth/ICT2.2/labwork/3/C++/automobile.cc
#include <regex>
#include <stdexcept>
#include <string>

#include "automobile.h"

using namespace std;

const regex license_pattern ("[0-9A-Z]+");

void
Automobile::set_license (string s)
{
  smatch m;
  if (!regex_match (s, m, license_pattern))
    throw invalid_argument{"invalid license plate"};
  license = s;
}

void
Automobile::set_fuel (double x)
{
  if (x < 0)
    throw invalid_argument{"negative fuel"};
  fuel = x;
}

void
Automobile::set_speed (double x)
{
  speed = (x < 0) ? 0 : x;
}

Automobile::Automobile (string l, double f, double s)
{
  set_license (l);
  set_fuel (f);
  set_speed (s);
}

void
Automobile::accelerate (double v)
{
  if (v < 0)
    throw invalid_argument{"negative acceleration"};
  if (fuel)
    set_speed (speed + v);
}

void
Automobile::decelerate (double v)
{
  if (v < 0)
    throw invalid_argument{"negative deceleration"};
  set_speed (speed - v);
}

File: usth/ICT2.2/labwork/3/Java/Automobile.java
import java.util.regex.Pattern;

class Automobile
{
  private static final Pattern licensePattern = Pattern.compile("[0-9A-Z]+");
  private double fuel;
  private double speed;
  private String license;

  public double getFuel()
  {
    return fuel;
  }

  public double getSpeed()
  {
    return speed;
  }

  public String getLicense()
  {
    return license;
  }

  public void setFuel(double fuel)
  {
    if (fuel < 0)
      throw new IllegalArgumentException(
        "fuel must be nonnegative, instead got " + fuel);
    this.fuel = fuel;
  }

  public void setSpeed(double speed)
  {
    this.speed = Math.max(0, speed);
  }

  public void setLicense(String license)
  {
    if (!licensePattern.matcher(license).matches())
      throw new IllegalArgumentException("invalid license: " + license);
    this.license = license;
  }

  public Automobile(double f, double s, String l)
  {
    setFuel(f);
    setSpeed(s);
    setLicense(l);
  }

  public void accelerate(double v)
  {
    if (v < 0)
      throw new IllegalArgumentException(
        "acceleration must be nonnegative, instead got " + v);
    if (fuel > 0)
      setSpeed(speed + v);
  }

  public void decelerate(double v)
  {
    if (v < 0)
      throw new IllegalArgumentException(
        "deceleration must be nonnegative, instead got " + v);
    setSpeed(speed - v);
  }
}
